src/Middleware/HttpLogging/src/BufferingStream.cs

```diff
@@ -0,0 +1,313 @@
+                // Convert with a empty ReadOnlySequence and flush: true until we get completed: true.
+
+                // This should never infinite due to the contract for decoders.
+                while (true)
+                {
+                    if (completed)
+                    {
+                        return new string(bufferWriter.WrittenSpan);
+                    }
+                    else
+                    {
+                        EncodingExtensions.Convert(decoder, ReadOnlySequence<byte>.Empty, bufferWriter, flush: true, out charUsed, out completed);
+                    }
+                }
+            }
+            catch (DecoderFallbackException ex)
+            {
+                _logger.DecodeFailure(ex);
+                return "<Decoder failure>";
+            }
+            finally
+            {
+                Reset();
+            }
+        }
+
+        public void Advance(int bytes)
+        {
+            if ((uint)bytes > (uint)_tailMemory.Length)
+            {
+                ThrowArgumentOutOfRangeException(nameof(bytes));
+            }
+
+            _tailBytesBuffered += bytes;
+            _bytesBuffered += bytes;
+            _tailMemory = _tailMemory.Slice(bytes);
+        }
+
+        public Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            AllocateMemory(sizeHint);
+            return _tailMemory;
+        }
+
+        public Span<byte> GetSpan(int sizeHint = 0)
+        {
+            AllocateMemory(sizeHint);
+            return _tailMemory.Span;
+        }
+
+        private void AllocateMemory(int sizeHint)
+        {
+            if (_head == null)
+            {
+                // We need to allocate memory to write since nobody has written before
+                BufferSegment newSegment = AllocateSegment(sizeHint);
+
+                // Set all the pointers
+                _head = _tail = newSegment;
+                _tailBytesBuffered = 0;
+            }
+            else
+            {
+                int bytesLeftInBuffer = _tailMemory.Length;
+
+                if (bytesLeftInBuffer == 0 || bytesLeftInBuffer < sizeHint)
+                {
+                    Debug.Assert(_tail != null);
+
+                    if (_tailBytesBuffered > 0)
+                    {
+                        // Flush buffered data to the segment
+                        _tail.End += _tailBytesBuffered;
+                        _tailBytesBuffered = 0;
+                    }
+
+                    BufferSegment newSegment = AllocateSegment(sizeHint);
+
+                    _tail.SetNext(newSegment);
+                    _tail = newSegment;
+                }
+            }
+        }
+
+        private BufferSegment AllocateSegment(int sizeHint)
+        {
+            BufferSegment newSegment = CreateSegment();
+
+            // We can't use the recommended pool so use the ArrayPool
+            newSegment.SetOwnedMemory(ArrayPool<byte>.Shared.Rent(GetSegmentSize(sizeHint)));
```

Comment:
When does this get returned?

Reply:
Either after the string is logged or when the stream is disposed.